sql: TESTING_RELOCATE

RaduBerinde opened the pull request:
For testing, we use `SHOW TESTING_RANGES` and the new LogicTest configuration
infrastructure - we run `ranges` in a 5-node testcluster.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14497)
<!-- Reviewable:end -->

RaduBerinde committed:
parser: add syntax for TESTING_RELOCATE

RaduBerinde committed:
sql: TESTING_RELOCATE implementation

For testing, we use `SHOW TESTING_RANGES` and the new LogicTest configuration
infrastructure - we run `ranges` in a 5-node testcluster.

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> hott

---

Review status: 0 of 15 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/split_at.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoOrbj873m1uz-UrWL:-KgoOrbj873m1uz-UrWM:b5r0cd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/c016d08d5a75f71cbede64e345a2432523a89c28/pkg/sql/split_at.go#L33)):*
> ```Go
> )
> 
> func getRowKey(
> ```

consider passing args by value to indicate that nulls are no good. I don't think performance matters much here.

---

*[pkg/sql/split_at.go, line 215 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoSD-oq4OVzgR_4dbU:-KgoSD-oq4OVzgR_4dbV:b8tysdc) ([raw file](https://github.com/cockroachdb/cockroach/blob/c016d08d5a75f71cbede64e345a2432523a89c28/pkg/sql/split_at.go#L215)):*
> ```Go
> 		if !cols[i].Typ.Equivalent(desiredTypes[i]) {
> 			return nil, errors.Errorf(
> 				"TESTING_RELOCATE data column %d must be of type %s, not type %s",
> ```

if it's easy, include the index column name in this error

---

*[pkg/sql/split_at.go, line 268 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoUwscYGkhGwRFIiYe:-KgoUwsdS2MFy1PqXhCj:bsdgr1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/c016d08d5a75f71cbede64e345a2432523a89c28/pkg/sql/split_at.go#L268)):*
> ```Go
> }
> 
> func (n *relocateNode) Next(ctx context.Context) (bool, error) {
> ```

maybe hint that each call to next is going to relocare one range, corresponding to one row from `n.rows`

---

*[pkg/sql/split_at.go, line 321 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoXA8PJ1Iet_qRzT-x:-KgoXA8PJ1Iet_qRzT-y:b-hbb06w) ([raw file](https://github.com/cockroachdb/cockroach/blob/c016d08d5a75f71cbede64e345a2432523a89c28/pkg/sql/split_at.go#L321)):*
> ```Go
> 	}
> 
> 	// Step 1: Add any targets that are not already replicas.
> ```

nit: Add any stores that *don't already have a replica of the range"

---

*[pkg/sql/split_at.go, line 324 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoXuwDkIMccm0BgG4X:-KgoXuwDkIMccm0BgG4Y:bjn8zef) ([raw file](https://github.com/cockroachdb/cockroach/blob/c016d08d5a75f71cbede64e345a2432523a89c28/pkg/sql/split_at.go#L324)):*
> ```Go
> 	//
> 	// TODO(radu): we can't have multiple replicas on different stores on the same
> 	// node, which can lead to some odd corner cases where we would have to first
> ```

this comment is good but it does't make it clear if the corner case is handled or, if not, what will happen

---

*[pkg/sql/parser/split.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoNsgSyc8-5zUnBZnW:-KgoNsgSyc8-5zUnBZnX:b-em64lw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7328dc6a00ab9d9d299a0fdf2c80ddd51e829c/pkg/sql/parser/split.go#L48)):*
> ```Go
> // statement.
> type Relocate struct {
> 	Table *NormalizableTableName
> ```

I know there was some discussion somewhere about this patter of having both a `NormalizedTableName` and a `TableNameWithIndex`; don't know what the conclusion was. Are these both supposed to be filled? Maybe add a comment (same above in `Split`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-:-KgoYORpQqtFqHMGuN8j:bb72svq)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 16 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[pkg/sql/split_at.go, line 33 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoOrbj873m1uz-UrWL:-KgpCqrmsqYORJIjTIYC:bk38l8v) ([raw file](https://github.com/cockroachdb/cockroach/blob/c016d08d5a75f71cbede64e345a2432523a89c28/pkg/sql/split_at.go#L33)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

consider passing args by value to indicate that nulls are no good. I don't think performance matters much here.
</blockquote></details>

I don't think changing to pass-by-value for that reason is a good pattern in general. I added a comment instead.

---

*[pkg/sql/split_at.go, line 215 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoSD-oq4OVzgR_4dbU:-KgpGeQhDAHA7nW0nrMo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c016d08d5a75f71cbede64e345a2432523a89c28/pkg/sql/split_at.go#L215)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if it's easy, include the index column name in this error
</blockquote></details>

Done.

---

*[pkg/sql/split_at.go, line 268 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoUwscYGkhGwRFIiYe:-KgpGs96e1e2RHHM5yKZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c016d08d5a75f71cbede64e345a2432523a89c28/pkg/sql/split_at.go#L268)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

maybe hint that each call to next is going to relocare one range, corresponding to one row from `n.rows`
</blockquote></details>

Done.

---

*[pkg/sql/split_at.go, line 321 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoXA8PJ1Iet_qRzT-x:-KgpGy9WTwvI5hAsq3bF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c016d08d5a75f71cbede64e345a2432523a89c28/pkg/sql/split_at.go#L321)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: Add any stores that *don't already have a replica of the range"
</blockquote></details>

Done.

---

*[pkg/sql/split_at.go, line 324 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoXuwDkIMccm0BgG4X:-KgpH7k9jcloiQpLzRdR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c016d08d5a75f71cbede64e345a2432523a89c28/pkg/sql/split_at.go#L324)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this comment is good but it does't make it clear if the corner case is handled or, if not, what will happen
</blockquote></details>

Done.

---

*[pkg/sql/parser/split.go, line 48 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-KgoNsgSyc8-5zUnBZnW:-KgpH8wZg0jw2c9vvC8h:b-f8sf1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/7d7328dc6a00ab9d9d299a0fdf2c80ddd51e829c/pkg/sql/parser/split.go#L48)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I know there was some discussion somewhere about this patter of having both a `NormalizedTableName` and a `TableNameWithIndex`; don't know what the conclusion was. Are these both supposed to be filled? Maybe add a comment (same above in `Split`).
</blockquote></details>

Added a comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 16 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14497#-:-KgpYWWteUtXRL6FBua1:bnfp4nl)*
<!-- Sent from Reviewable.io -->
